### PR TITLE
#825: Allow passing explicit connection to `ACL.{reload,save}`

### PR DESCRIPTION
**Pull request by tseaver**
See #825

### CODE REVIEW

File: gcloud/storage/acl.py [hunk 1]
```diff
@@ -345,78 +346,84 @@ def get_entities(self):
         self._ensure_loaded()
         return list(self.entities.values())
 
-    def reload(self):
-        """Reload the ACL data from Cloud Storage.
+    @property
+    def connection(self):
+        """Compute the connection for API requests for this ACL.
```
Comment (dhermes) [inline gcloud/storage/acl.py]: Why introduce this property if the end goal of #825 is to unbind connection from objects?

Also, docstring still indicates this is a method, but it is a property.

File: gcloud/storage/acl.py [hunk 2]
```diff
@@ -345,78 +346,84 @@ def get_entities(self):
...
 
         This is a virtual method, expected to be implemented by subclasses.
 
-        :raises: :class:`NotImplementedError`
+        :raises: :class:`NotImplementedError` if ``_connection`` attribute
+                 is not set on the instance.
+
         """
+        # Allow override for testing
+        connection = getattr(self, '_connection', None)
```
Comment (dhermes) [inline gcloud/storage/acl.py]: Why is this necessary? Why can't we just set `_connection` on test objects?

File: gcloud/storage/acl.py [hunk 3]
```diff
@@ -345,78 +346,84 @@ def get_entities(self):
...
+        """Reload the ACL data from Cloud Storage.
+
+        :type connection: :class:`gcloud.storage.connection.Connection` or None
+        :param connection: explicit connection to use for API request;
+                           defaults to instance property.
         """
-        super(BucketACL, self).__init__()
-        self.bucket = bucket
+        if connection is None:
+            connection = self.connection
```
Comment (dhermes) [inline gcloud/storage/acl.py]: Same comment here as I had in #852, this is not the same spirit as the changes in #843.

File: gcloud/storage/acl.py (hunk 2 above)
Comment (tseaver) [inline gcloud/storage/acl.py]: Because this PR wasn't aimed at removing `self.connection` usage?  Anyway, having the base class define a property which raises unconditionally makes it impossible to test the "hoisted" methods (which were entirely redundant, except for paths).

Comment (dhermes) [inline gcloud/storage/acl.py]: I'm unclear on what was impossible to test. Can you provide a snippet?

Comment (tseaver) [inline gcloud/storage/acl.py]: This PR moves the implementations of `reload`, `save`, and `clear` up to the base `ACL` class.  Without having some means of setting a connection, it is not possible to test that the default case (no connection passed) works correctly.

I just checked:  #843 actually fails to test the "no connection passed" case for most methods.

Comment (dhermes) [inline gcloud/storage/acl.py]: ISTM we would end up with 2x as many tests when in reality testing `_require_connection` should be sufficient.

**Comment by tseaver** [comment]
@dhermes PTAL

File: gcloud/storage/acl.py [hunk 4]
```diff
@@ -345,77 +348,64 @@ def get_entities(self):
...
+        """Compute the path for GET API requests for this ACL.
 
         This is a virtual method, expected to be implemented by subclasses.
 
-        :raises: :class:`NotImplementedError`
+        :raises: :class:`NotImplementedError` if ``_reload_path`` attribute
+                 is not set on the instance.
         """
+        # Allow override for testing
+        path = getattr(self, '_reload_path', None)
```
Comment (dhermes) [inline gcloud/storage/acl.py]: I'm still hungup on why this is necessary.

File: gcloud/storage/acl.py [hunk 5]
```diff
@@ -441,21 +434,38 @@ def clear(self):
...
+        """
+        :type bucket: :class:`gcloud.storage.bucket.Bucket`
+        :param bucket: The bucket to which this ACL relates.
         """
-        self.save([])
+        super(BucketACL, self).__init__()
+        self.bucket = bucket
+
+    @property
+    def connection(self):
```
Comment (dhermes) [inline gcloud/storage/acl.py]: Is this still needed after the virtual `connection` was axed?

File: gcloud/storage/acl.py (hunk 4 above)
Comment (tseaver) [inline gcloud/storage/acl.py]: The hoisted `reload` implementation in the base ACL class relies on having a `reload_path` property, which is implemented via a "pure virtual" method in the base class (raising `NotImplementedError`).  We cannot write a test for the `reload` implementation in that class (it *used* to be pure virtual too, but now uses the property).  We would either need to subclass to test it, or else allow tests to inject a value which overrides the pure virtualness.

Comment (dhermes) [inline gcloud/storage/acl.py]: And why are either of those options too hard? I choose either over the `getattr` bits above.

Comment (tseaver) [inline gcloud/storage/acl.py]: The code you are objecting to *is* the second option -- it allows the base `ACL` class to have a "real" `reload_path` value for testing purposes:  we can't assign to the property (because it is read only), but we can assign to another attribute.  The only other alternative would be to remove the "pure virtual" bit:  maybe a `reload_property = None` assigned at class scope, which could be overridden for testing purposes on an instance.

Comment (dhermes) [inline gcloud/storage/acl.py]: Why can't the test just overwrite `reload_path`?

**Comment by tseaver** [comment]
@dhermes d4bc450 drops the "pure virtual" property implementation of `reload_path` and `save_path`, replacing them with a class attr default of None, allowing tests to assign to the `ACL` instances.  Any remaining issues?

**Comment by dhermes** [comment]
LGTM.

@tseaver I'd like to have a pow-wow about the current state of our ACL support and try to figure out what we want (or better yet, what someone wants that's actually used ACLs for an application). Things are a bit out of place in the current state (e.g. `roles` is a set, but storage has a concept of concentric roles for the very purpose of avoiding an ACL having more than one role).

**Comment by tseaver** [comment]
It is annyoing that those docs can't keep role vs. permisison straight, even on the same page.

If `READER`, `WRITER`, and `OWNER` are the only possible permissions / roles, then we could avoid having an ACE (a single assertion/entity, of which an ACL is a collection) have only one of them.

**Comment by tseaver** [comment]
We should also work out how to support the [predefined ACLs feature](https://cloud.google.com/storage/docs/access-control#predefined-acl)
